content/chapters/compute/lab/content/user-level-threads.md

```diff
@@ -0,0 +1,87 @@
+To compile and run the program, do the following steps:
+
+```console
+cd build/
+cmake -S .. -B .
+make
+./simple
+```
+
+The `cmake` step must be executed only once.
+After modifying the source files, it is enough to run `make`.
+
+## No system calls
+
+Use `strace` to find calls to `clone()` in the execution of `simple`.
+Can you find any?
+Remember that `clone()` is the system call used to create threads, as pointed out [here](./arena.md#threads-and-processes-clone).
+
+## Synchronization
+
+By default, the fibers that run on the same thread are synchronized - no race-conditions can occur.
+This is illustrated by the `support/user-level-threads/sum.cc` implementation.
+
+The user can, however, implement further synchronization, by using the `yield()` call, or classic synchronization methods, like mutexes, barriers and condition variables.
+
+### Yielding
+
+As the scheduler is a cooperative one, each fiber can yield (or not), to allow another fiber to run.
+Follow the `support/user-level-threads/yield_launch.cc` implementation and run it.
+Note the `boost::fibers::launch::dispatch` parameter provided to the fiber constructor.
+It notifies the scheduler to start the fibre as soon as it is created.
+In order to explain the output, we must consider that the fibers are created by a **main fiber**, that is scheduled along with the others, in this case.
+
+Modify the launch parameter into `boost::fibers::launch::post`, compile and notice the differences.
+The `post` parameter notifies the scheduler not to start the fibers imediately, but rather place them into an execution queue.
+Their execution will start after the main fiber calls the `join()` function.
+
+### Barriers
+
+Follow the `support/user-level-threads/yield_barrier.cc` implementation.
+It uses a barrier to achieve the same result as the previos implementation, that used `post` as the launch parameter.
+
+## Interaction Between Threads and Fibers
+
+As we mentioned before, multiple fibers can run on the same thread, and a scheduler is implemented on each thread.
+By default, the scheduler is `round_robin`.
+It runs the fibers, in the order of their creation, until they yield or finish their work.
+If a fiber yields, it is placed at the back of the round-robin queue.
+Using this scheduler, each thread uses only its fibers;
+if one thread has more work to do than another, bad luck.
+
+But there are other scheduler implementations, such as `shared_work` and `work_stealing`.
+Follow the `support/user-level-threads/threads_and_fibers.cc` implementation.
+It creates multiple fibers and threads, and uses the `shared_work` scheduler to balance the workload between the threads.
+Each main fiber, from each thread, is suspended until all worker fibers have completed their work, using a condition variable.
+
+```cpp
+cnd_count.wait( lk, [](){ return 0 == fiber_count; } );
+```
+
+It also uses `thread local storage` and `fiber local storage` to store the ID of each thread / fiber.
```

Comment:
Which is "It"? Elaborate on this a bit. If it's not present in `threads_and_fibers.cc`, then skip it.

Reply:
it, the program

Reply:
Still vague. Use a proper subject.